add blog post announcing proposed changes/oteps for stability work (#8208)
open-telemetry/opentelemetry.io: content/en/blog/2025/stability-proposal-announcement.md

and maturity of the project.

## Improving defaults for stable components

The first proposal has to do with how SIGs and the project communicate stability
of components – such as instrumentation libraries, Collector receivers, or tools
like weaver. We will clearly state on the project website which components are
stable. We may list un-stable components as well, but the main goal is to
provide end-users with a single source of truth about what they can expect to be
stable. Stability requirements will also be streamlined and clarified for
contributors, maintainers, and end-users. We plan to expand these stability
requirements to include documentation, examples, implementation scenarios,
performance benchmarks, and other useful information.

In addition, we are proposing a change to the defaults of the SDK and
instrumentation agents. By default, stable components should not automatically
import, include, or enable unstable components. For example, if you install a
stable distribution of OpenTelemetry Java Instrumentation Agent, the default
receivers, processors, samplers, exporters, and libraries should not include

> alpha or experimental instrumentations (which is the current behavior).

Reviewer: Not only is this the current behavior for the Java Instrumentation Agent, this is also the a[pproach that we announced that we would be following for Android](https://opentelemetry.io/blog/2025/android-road-to-stable/#looking-ahead). This change injects chaos into that plan, so we will need to pause this plan until this is figured out.

Author: well, if you're planning on doing what the java agent does, then I think you'd be fine...?

Reviewer 2: oh, sorry, @austinlparker I think I got our wires crossed when I answered your question the other day

the Java agent current behavior _does_ include unstable things by default

Committee.

OpenTelemetry is a large, complex, and highly involved project. A quick glance
at the
[specification compliance matrix](https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md)
should be enough to tell you that supporting tracing, metrics, logs, and
profiles across a dozen languages – not to mention the complexity that comes
from different language runtimes and execution environments – means that there’s
a _lot_ to take in. One thing we’ve heard from you over the past months is that
this complexity makes it difficult to get started, or get going, with
OpenTelemetry as an end user. It also makes life difficult for our maintainers
and contributors. How should they prioritize feature development versus spec
changes versus instrumentation stability?

With this in mind, the OpenTelemetry project would like to announce some
proposed changes that we’d like to institute, with the goal of normalizing our
language around component stability and component offerings.

We believe that these changes need community involvement and discussion to be a
success, so we’re taking this opportunity to announce our intention and open an

> [issue for discussion](https://github.com/open-telemetry/community/issues/3086)

Reviewer 3: Maybe an actual "discussion" (though they are disabled on the community repo) would make more sense? It allows threaded comments, which makes it easier to follow and answer specific comments.

Reviewer 4: We had a bit of back-and-forth about what would be more appropriate, to be honest my preference would be to have multiple issues

Author: I think there's pros and cons to both, but I'd tend to agree with @dmathieu that a discussion would make more sense thanks to its threaded nature. multiple issues seem like they'd be kinda gnarly for people to read thru/get the full discussion context.

in order to get feedback from users, maintainers, and contributors. We do not
anticipate that these changes will be completed overnight, and want to assure
everyone that we will continue to prioritize our existing commitments to users
and maintainers even as we consider necessary changes for the overall wellbeing
and maturity of the project.

> ## Improving defaults for stable components

Reviewer 4: The "defaults" is the part that I think can be confusing, actually: it sounds like you are talking about the default features on the components instead of the default components that are available

> ## Improving defaults for stable components

Reviewer 4: "stable defaults" is not something we have aimed for in the Collector: if it makes sense to change a default and it does not lead to breakage from users we will change it. I am bit confused as to why that is important?

> ## Improving defaults for stable components

Author: I've clarified the stability intentions a bit more and simplified how it's communicated a bit; ptal

The first proposal has to do with how SIGs and the project communicate stability
of components – such as instrumentation libraries, Collector receivers, or tools

> like weaver. We will clearly state on the project website which components are

Reviewer 5: do you mean, we will add stability information to the OpenTelemetry registry, and allow people to filter by stability?

stable. We may list un-stable components as well, but the main goal is to
provide end-users with a single source of truth about what they can expect to be
stable. Stability requirements will also be streamlined and clarified for
contributors, maintainers, and end-users. We plan to expand these stability

> requirements to include documentation, examples, implementation scenarios,

Reviewer 5: maybe a link to an issue to explore this further?

performance benchmarks, and other useful information.

In addition, we are proposing a change to the defaults of the SDK and
instrumentation agents. By default, stable components should not automatically
import, include, or enable unstable components. For example, if you install a
stable distribution of OpenTelemetry Java Instrumentation Agent, the default
receivers, processors, samplers, exporters, and libraries should not include
alpha or experimental instrumentations.

SSimilarly, the OpenTelemetry JavaScript Node Automatic Instrumentation

> Metapackage would only include stable instrumentations (e.g., those which are at

Reviewer 6: I should've better marked this as comment as an FYI for the other JS Maintainers that we should probably focus our efforts toward stability of these packages ASAP - it's not intended as a blocker for this PR.

Stabilization has mostly been a prioritization problem for us. This blog post helps, IMO.

Stability for these base packages can be achieved - like you said - by putting unstable features behind feature gates (or experimental entrypoints, in our case, most likely).

Second, we are adjusting our overall posture when it comes to instrumentation
and instrumentation stability. Our goal, as a project, has always been to see
other libraries, frameworks, runtimes, and tools
[natively adopt OpenTelemetry](https://www.youtube.com/watch?v=l8xiNOCIdLY).
This is still the goal – we want other developers to embed our APIs into their
software, rather than requiring you to download and install libraries that
inject instrumentation code into third party dependencies. However, many users
have come to see our ‘contrib’ repositories and distributions as generic
replacements for system telemetry agents.

We’re thrilled that we’ve had so many users adopt the Collector as a
vendor-agnostic solution for collecting telemetry data, and we believe that it’s
a strong testament to the long-term viability of the project. What has become
challenging is that our contrib repositories – especially in the Collector –
represent a huge amount of work for maintainers and a constant drag on project
velocity. Even in a world where we had twice as many contributors and
maintainers, it would still be a challenge to accept all desired receivers or

> instrumentation libraries and ensure they all met a sufficiently high quality

Reviewer 5: I'm not sure if instrumentation libraries should be related to the collector work. I might be wrong about this, but it doesn't seem like the collector expertise transfers to the instrumentation libraries, ie, a person able to make a mysql receiver is not actually the best at building a python instrumentation library, even on mysql.

bar for release.

Our next proposal is designed to address this tension by more tightly scoping
what instrumentation libraries and receivers we’ll consider for inclusion and
stability. In short, instrumentation should be an implementation of a semantic
convention. This has quite a few implications, some of which include:

- Existing instrumentation libraries or Collector receivers that do not follow
  an established semantic convention will be treated as experimental/unstable

>   until such time those conventions exist.

Reviewer 5: exist, as in, accepted and merged in the specification repository? Or OK to have a weaver model in a gist somewhere?

Reviewer 7: For future readers of the blogpost, I think it would be good to clarify here that by "established" we don't mean "stable" semconv. Stable instrumentation libs and receivers could have breaking changes in terms of the telemetry emitted if the underlying semconv is not stable. We should explicitly call that out in versioning as mentioned below, i.e. "stable (beta)".

Author: I've changed this entire section around to be more clear about intent without over indexing on details like this, ptal.

- SIGs should no longer accept instrumentation libraries or receivers that do
  not implement a semantic convention.
- The priority for instrumentations should be driven by semantic convention
  stability efforts.

Additionally, we will recommend that instrumentation which depends on a beta

> version of a semantic convention may be marked as stable (beta), and included in

Reviewer 8: I really like this about following semantic conventions but I can see this becoming a major roadblock given latest semconv triage process.

What I think would be worth mentioning near this section is that Semantic conventions will have a defined process including requirements to be satisfied for an attribute/signal to promoted to a new stability level.

Note currently alpha & beta levels are not used in semconv and instead it goes experimental to rc.

Author: I've simplified the way we talk about this a bit to be more clear about the intent and some design requirements around semconv, hopefully that addresses this feedback

> version of a semantic convention may be marked as stable (beta), and included in

Reviewer 5: Is beta stable? We explicitly have beta and stable separate concepts in the collector stability levels.

Author: tried to capture this in the post better, ptal

default distributions. The goal here is to reduce the dwell time of
instrumentation libraries in experimental maturity phases; We believe that
sufficient tooling exists via schema transformations or processors, as well as
the dual-write convention for updated attributes, as to allow end users to
depend upon beta conventions.

This change is not meant to ‘yank away’ integrations that current users may
depend on, but to more clearly communicate the expected future state of various
components, to provide accurate guidance to end users about the potential for
instrumentation to change, and for contributors and maintainers to have a
significantly more transparent and simplified roadmap of where to focus efforts
around stability work.

> ## Unified releases across the project

Reviewer 9: Is this a proposal or a direction?

FWIW - I think this is a directional/shift and decision we're making.

The proposal should follow this with details on how we'll accomplish this, but for a blog post calling all of these "proposals" doesn't really sit with me as a "Ok, OpenTelemetry is going to change how it does things".

I'd like to use *stronger* language, and link to proposals that are evolving for details.

> ## Unified releases across the project

Reviewer 4: > I'd like to use stronger language, and link to proposals that are evolving for details.

What does stronger language look like here?

> ## Unified releases across the project

Reviewer 9: An example, instead of 

```
Our final proposal is to normalize release cycles for the project as a whole,
```

We would have something like:

```
Finally, we are working a project-wide release cycle, similar to how projects like Kubernetes function.
```

It's a minor difference, but an important one.  A proposal may not be accepted, but this is suggested we are kicking off a task / project for the work.

If we're posting a public blog, the uncertainty of a proposal seems dubious to me.  We're effectively both claiming its a proposal, but advertising it in a way we can't really say 'no' going forward.

I'd rather us definitely move a direction and tell people in a public blog (encouraging accountability to the direction), or instead use the existing OpenTelemetry proposal process if this is just a proposal phase.

> ## Unified releases across the project

Reviewer 4: I am fine with that, but I am also scared about overpromising before examining what is feasible with the community. I think in your example it is fine to me, but other parts (e.g. you talk in another comment about semantic conventions burden with some of the other changes) may not be feasible to word as strongly without overpromising

---

<!-- If your post doesn't have secondary authors, then delete the following paragraph: -->

With contributions from, and on behalf of, the OpenTelemetry Governance
Committee.

OpenTelemetry is a large, complex, and highly involved project. A quick glance
at the
[specification compliance matrix](https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md)
should be enough to tell you that supporting tracing, metrics, logs, and
profiles across a dozen languages – not to mention the complexity that comes
from different language runtimes and execution environments – means that there’s
a _lot_ to take in. One thing we’ve heard from you over the past months is that
this complexity makes it difficult to get started, or get going, with
OpenTelemetry as an end user. It also makes life difficult for our maintainers
and contributors. How should they prioritize feature development versus spec
changes versus instrumentation stability?

With this in mind, the OpenTelemetry project would like to announce some
proposed changes that we’d like to institute, with the goal of normalizing our
language around component stability and component offerings.

> We believe that these changes need community involvement and discussion to be a

Reviewer 10: who's "we" here? GC? GC and TC? Maintainers? Make it clear to the reader.

and maturity of the project.

## Improving defaults for stable components

The first proposal has to do with how SIGs and the project communicate stability
of components – such as instrumentation libraries, Collector receivers, or tools
like weaver. We will clearly state on the project website which components are
stable. We may list un-stable components as well, but the main goal is to
provide end-users with a single source of truth about what they can expect to be
stable. Stability requirements will also be streamlined and clarified for
contributors, maintainers, and end-users. We plan to expand these stability
requirements to include documentation, examples, implementation scenarios,
performance benchmarks, and other useful information.

In addition, we are proposing a change to the defaults of SDKs and
instrumentation agents. By default, stable components should not automatically
import, include, or enable unstable components. For example, if you install a
stable distribution of OpenTelemetry Java Instrumentation Agent, the default
receivers, processors, samplers, exporters, and libraries should not include

> alpha or experimental instrumentations.

Author: Clarified the current thinking around the default stability level in the post; ptal

Similarly, the OpenTelemetry JavaScript Node Automatic Instrumentation
Metapackage would only include stable instrumentations (i.e., those which are at
v1.0 or greater). For consistency, we propose a single project-wide way to

> enable unstable behavior for end-users. The goal of this proposal is to reduce

Reviewer 7: I think we should mention here that this is something that we expect to be in the spec, i.e. not a blogpost or some general comms.

Author: added language to this effect where appropriate

vendor-agnostic solution for collecting telemetry data, and we believe that it’s
a strong testament to the long-term viability of the project. What has become
challenging is that our contrib repositories – especially in the Collector –
represent a huge amount of work for maintainers and a constant drag on project
velocity. Even in a world where we had twice as many contributors and
maintainers, it would still be a challenge to accept all desired receivers or
instrumentation libraries and ensure they all met a sufficiently high quality
bar for release.

Our next proposal is designed to address this tension by more tightly scoping
what instrumentation libraries and receivers we’ll consider for inclusion and
stability. In short, instrumentation should be an implementation of a semantic
convention. This has quite a few implications, some of which include:

- Existing instrumentation libraries or Collector receivers that do not follow
  an established semantic convention will be treated as experimental/unstable
  until such time those conventions exist.
- SIGs should no longer accept instrumentation libraries or receivers that do
  not implement a semantic convention.

> - The priority for instrumentations should be driven by semantic convention

Reviewer 9: For focusing the community and making it clear what we want to drive I like this.

I"m worries this will put pressure to start up N semantic convention groups that don't exist today.  We should make sure our Project onboarding process and traige can handle this new pressure.

cc @open-telemetry/specs-semconv-maintainers

cross-compare an array of version tags or package management files, our goal
here would be to have a simple meta-target that could be rolled out across an
organization with consistent documentation, features, etc.

The goal of this is to improve the ability of external organizations to
interpret OpenTelemetry release cycles and support. It paves the way to official
LTS releases in the future, and ideally serves as a more stable target for
downstream integrations and service providers – rather than potentially having
to keep pace with dozens of releases a month, you can focus on just the ‘major’
releases.

## What's next after this?

Looking forward, our goal with these changes is to not simply change for the
sake of change; These are a recognition of the impact and importance of
OpenTelemetry to the cloud native software community.
[OpenTelemetry has been the second highest velocity project in the CNCF over the past few years](https://www.cncf.io/wp-content/uploads/2025/04/CNCF-Annual-Report-2024_v2.pdf),
and
[nearly 50% of surveyed cloud native end user companies have adopted the project](https://www.cncf.io/wp-content/uploads/2025/04/cncf_annual_survey24_031225a.pdf).

> The changes we are proposing here are designed to set us up for even greater

Author: at this point pretty much all details have been removed from the post. if there's something you think is still over specified please point it out directly.

- User interviews, surveys, and discussions have demonstrated that the
  complexity of OpenTelemetry creates impediments to production deployments.

**What's Changing?**

- We'll be moving to a simpler, but stricter, definition of what components can
  be present in a stable release and updating stability requirements to include
  new guidance on documentation, performance testing, etc.
- Semantic convention guidelines will be updated to encourage faster iteration,
  greater federation, and to unblock consumers to allow stable instrumentation
  releases.
- To make these releases easier to consume, we'll be introducing 'epoch
  releases' -- similar to the Kubernetes release cycle.

**What we need your input on:**

- Implementation timelines and stability requirements.
- Proposed changes to semantic convention stability.
- How we can make releases better.

[Join the discussion!](https://github.com/open-telemetry/community/discussions/3098).

## Why are we doing this?

OpenTelemetry is a large, complex, and highly involved project. A quick glance
at the
[specification compliance matrix](https://github.com/open-telemetry/opentelemetry-specification/blob/main/spec-compliance-matrix.md)

> should be enough to tell you that supporting tracing, metrics, logs, and

Reviewer 11: This para should be re-written. "should be enough to tell you" reads a bit odd - I suggest changing to "A quick glance at the specification <link> shows you that supporting tracing, metrics, logs and profiles across different languages and environments is complex".

profiles across a dozen languages – not to mention the complexity that comes
from different language runtimes and execution environments – means that there’s
a _lot_ to take in.

For end-users, this complexity holds back adoption and introduces surprise.
We've heard feedback about components whose config changes out from under them,
performance regressions when taking minor updates, and difficulty managing
rollouts at scale.

For maintainers, this complexity makes their job harder than it needs to be.
There's a lack of clear milestones and guidance about what's 'most important' at

> any given time. Stability efforts involve a lot of whiplash and there's often

Reviewer 11: Can "whiplash" be changed to "churn"?

Most of the value our community derives from OpenTelemetry comes directly from
instrumentation libraries and Collector components – not the core SDKs. While we
organize these as `contrib` repositories to distinguish them from core
components, end users don't see or care about this distinction. They just want
instrumentation that works.

For maintainers and project leadership, our stability goals and the nature of
`contrib` present a significant challenge. Users want stable, well-tested, and
performant releases -- that _also_ perform the same function as commercial
instrumentation agents.

### What's Changing?

There's three major changes that the project will be rolling out. Keep in mind
the exact details of some of this may change, but these goals are more or less
fixed.

1. All components across all repositories should adhere to a consistent way of
   communicating stability, through a metadata file, that can be discovered and
   parsed in a programmatic way. The exact format will be defined through an
   OTEP and incorporated into the specification. As part of this, we'll be
   normalizing stability levels across components, including semantic

>    conventions by introducing alpha/beta stability to that effort.

Author: I've removed the specific alpha/beta stuff and incorporated the decoupling language into the post.

2. Stability requirements will be expanded to include more requirements around
   documentation and where it's hosted, example code, performance benchmarks
   (where applicable), implementation cookbooks, and other artifacts as
   necessary.
3. Stable components must only enable other stable components by default. A new
   global configuration option shall be introduced that allows adopters to
   choose a desired minimum stability level, with a default value of
   `[stable, rc, beta]`.

This is a big change for maintainers, especially those who have shipped v1+ of
their libraries. We would deeply appreciate your feedback on this proposal in

> the [discussion](https://github.com/open-telemetry/community/discussions/3098).

Author: given that this is probably a multi-threaded topic and issues don't work great for that, I think a discussion is a good venue for the overall conversation, especially since the individual OTEPs will be better at hashing out single-threaded details?

stability.

### Background

Semantic conventions evolve slowly and deliberately because they must work
across diverse telemetry systems. While OpenTelemetry is designed for
interconnected signals flowing together, users deploy many different storage and
analysis engines to consume this data. Each backend has its own constraints and
capabilities. Maintainers must balance competing concerns – keeping cardinality
manageable, ensuring attributes are useful but not overly specific, and making
conventions that work well regardless of where the data ends up.

The downside of this is that progress on semantic conventions can be slow, and
this slowness impacts all consumers of the conventions. Many instrumentation
libraries are currently stuck on pre-release versions because they depend on
experimental semantic conventions. Outside contributors are stuck between
emitting unspecified telemetry or trying to engage in the process, which
requires a long commitment. Finally, we're internally inconsistent in
instrumentation across the project; Some libraries are mapped to conventions,
others exist independently of it.

### What's Changing?

Our goals here are designed to achieve two outcomes.

> 1. Semantic conventions should have more nuanced stability levels in order to

Author: pushed some changes here, ptal

   unblock upstream development and gain real world feedback from users.

> 2. Semantic conventions should be more federated; We don't want the project to

Reviewer 11: "We" is confusing again. Is it the project maintainers or GC or TC or all of the above?

should prioritize instrumentations that align to existing conventions and
deprioritize others. Second, we would like to promote more first-party semantic
conventions to 'beta' stability in order to unblock downstream instrumentation
libraries. Many of our conventions have months, or years, of real-world use that
we can rely on as guidance for their fitness. Finally, we'd like to formalize
the semantic convention format specification to the point that third-parties
outside the OpenTelemetry project feel comfortable using it to publish their
own, federated, semantic conventions.

In order to accomplish this, we're looking for feedback on several areas from
maintainers and end-users -- especially around the maturity/lifecycle of
semantic conventions, as well as what's missing in terms of federating semantic
conventions. We are more flexible on proposals here, but our outcomes aren't.
Remember, a core goal of the project is to encourage other libraries, tools, and
frameworks to
[natively adopt OpenTelemetry](https://www.youtube.com/watch?v=l8xiNOCIdLY) --
semantic conventions are a big part of that.

## 3. Confident, Stable, Releases.

## Background

> OpenTelemetry isn't just one thing, especially in complex deployments. Subtle

Reviewer 11: Suggest re-phrasing. "one thing" may reduce the importance of what is being conveyed. Maybe we could say "OpenTelemetry is not just one component being developed and released for complex deployments."

performant releases -- that _also_ perform the same function as commercial
instrumentation agents.

### What's Changing?

There's three major changes that we'll be rolling out. Keep in mind the exact
details of some of this may change, but these goals are more or less fixed.

1. All components across all repositories should adhere to a consistent way of
   communicating stability, through a metadata file, that can be discovered and
   parsed in a programmatic way. The exact format will be defined through an
   OTEP and incorporated into the specification. As part of this, we'll be
   normalizing stability levels across components, including semantic
   conventions by introducing alpha/beta stability to that effort.
2. Stability requirements will be expanded to include more requirements around
   documentation and where it's hosted, example code, performance benchmarks
   (where applicable), implementation cookbooks, and other artifacts as
   necessary.
3. Stable components must only include other stable components by default. A new
   global configuration option shall be introduced that allows adopters to
   choose a desired minimum stability level, with a default value of

>    `[stable, rc, beta]`.

Reviewer 4: We can leave that part for the OTEP
```suggestion
   choose a desired minimum stability level.
```

Author: Hm, I think we want to commit to 'beta' being the default stability in this piece since we're also proposing adding beta stability for semconv?

Reviewer 4: Okay, fair

Reviewer 9: I think picking specific names and meaning for stability level is best done the OTEP.  Best message here is "we'll have a default value that is both stable and includes components you need for a complete solution".

issue: the issue ID for this blog post # TODO: See https://opentelemetry.io/docs/contributing/blog/ for details
sig: Governance Committee # TODO: add the name of the SIG that sponsors this blog post
---

## Summary

OpenTelemetry is changing how the project approaches stability and releases,
across all sub-projects and repositories. We would like to make a clearer
distinction between internal and external stability levels; As a project, we
should be more concerned with the packaging and 'product' of OpenTelemetry
distributions.

**Why?**

- User interviews, surveys, and discussions have demonstrated that the
  complexity of OpenTelemetry creates impediments to production deployments.

**What's Changing?**

- We'll be moving to a simpler, but stricter, definition of what components can
  be present in a stable release and updating stability requirements to include
  new guidance on documentation, performance testing, etc.
- Semantic convention guidelines will be updated to encourage faster iteration,
  greater federation, and to unblock consumers to allow stable instrumentation

>   releases.

Reviewer 2: ```suggestion
- Instrumentation stability guidelines will be updated to encourage faster iteration,
  greater federation, and to unblock consumers to allow stable instrumentation
  releases.
```

- To make these releases easier to consume, we'll be introducing 'epoch
  releases' -- similar to the Kubernetes release cycle.

> **What we need your input on:**

Author: did you read the rest of it? it says in each individual section what we're looking for feedback on.

methods and modes, existing tools and strategies, which means we need to provide
implementations of not just the specification, but _applications_ of it as well.
In practice, this means we need to distribute libraries to add OpenTelemetry
instrumentation to existing HTTP servers and clients, or Collector receivers to
scrape metrics from MySQL and translate them into OTLP.

Most of the value our community derives from OpenTelemetry comes directly from
instrumentation libraries and Collector components – not the core SDKs. While we
organize these as `contrib` repositories to distinguish them from core
components, end users don't see or care about this distinction. They just want
instrumentation that works.

For maintainers and project leadership, our stability goals and the nature of
`contrib` present a significant challenge. Users want stable, well-tested, and
performant releases -- that _also_ perform the same function as commercial
instrumentation agents.

### What's Changing?

There's three major changes that the project will be rolling out. Keep in mind
the exact details of some of this may change, but these goals are more or less
fixed.

> 1. All components across all repositories (including semantic conventions)

Author: It's relevant to end users? People don't understand how we communicate stability currently. This is especially pernicious for ecosystem adopters outside of the otel org; There's no guidance on how they should interoperate with our stability guidelines so they just kinda do whatever.

   should adhere to a consistent way of communicating stability, through a
   metadata file/information, that can be discovered and parsed in a
   programmatic way. The exact format will be defined through an OTEP and
   incorporated into the specification.
2. Stability requirements will be expanded to include more requirements around
   documentation and where it's hosted, example code, performance benchmarks
   (where applicable), implementation cookbooks, and other artifacts as
   necessary.
3. Stable distributions of OpenTelemetry should only enable stable components by
   default. Users should be able to select a desired minimum stability level
   with a documented and consistent configuration option.

This is a big change for maintainers, especially those who have shipped v1+ of
their libraries. We would deeply appreciate your feedback on this proposal in
the [discussion](https://github.com/open-telemetry/community/discussions/3098).

## 2. Instrumentation Stability and Semantic Conventions

> As mentioned above, we would like to adjust the semantic convention stability

Reviewer 2: ```suggestion
As mentioned above, we would like to adjust the instrumentation stability
```

stability.

### Background

Semantic conventions evolve slowly and deliberately because they must work
across diverse telemetry systems. While OpenTelemetry is designed for
interconnected signals flowing together, users deploy many different storage and
analysis engines to consume this data. Each backend has its own constraints and
capabilities. Maintainers must balance competing concerns – keeping cardinality
manageable, ensuring attributes are useful but not overly specific, and making
conventions that work well regardless of where the data ends up.

The downside of this is that progress on semantic conventions can be slow, and
this slowness impacts all consumers of the conventions. Many instrumentation
libraries are currently stuck on pre-release versions because they depend on
experimental semantic conventions. Outside contributors are stuck between
emitting unspecified telemetry or trying to engage in the process, which
requires a long commitment. Finally, we're internally inconsistent in
instrumentation across the project; Some libraries are mapped to conventions,
others exist independently of it.

### What's Changing?

Our goals here are designed to achieve three outcomes.

> 1. Instrumentation stability should be decoupled from semantic convention

Reviewer 12: I am not sure if this is a good point for end users. The end users depend on semantic conventions in their visualization and alerting tools. 

As a user updating the instrumentation library, I am very interested in knowing if the instrumentation changed any attributes, because it can break my dashboarding and alerting.

Reviewer 12: Shall it be mentioned in this document?

>    stability.

Reviewer 2: ```suggestion
1. Instrumentation stability should be decoupled from semantic convention
   stability.
```

2. Semantic conventions should be more federated; OpenTelemetry should not be
   the final word on what conventions exist, and instead should focus on
   creating core conventions that can be extended and built upon.
3. Semantic convention development and iteration should not be a blocker on
   distribution maintainers.

To this end, we have a few recommendations we'd like to codify into the
specification. First, our position around instrumentation libraries in
OpenTelemetry is that they exist as concrete implementations of the semantic
conventions. This gives us a concrete target for 'first party' instrumentation
libraries that we wish to support in distributions. In addition, maintainers
should prioritize instrumentations that align to existing conventions and
deprioritize others.

Second, we'd like to make it easier for maintainers to ship stable
instrumentations. If an instrumentation's API surface is stable, then we believe
that semantic convention stability should not block the stabilization of that
instrumentation library. This means that we'll need to be thoughtful in

> providing migration pathways for telemetry as operators upgrade to new major versions

Reviewer 3: That makes sense. Then, how about something like

```
This means we'll need to be thoughtful in providing ways to safely use unstable versions, and migrate when breaking changes happened
```

Collector. Enterprise deployments and upgrades are slow, deliberate things --
teams simply do not have the bandwidth to validate and roll out changes at the
cadence we ship.

## Goals and Objectives

Ultimately, our goal here is to make it easier for large organizations to deploy
OpenTelemetry. Please keep in mind that in many organizations, 'deployment' and
'upgrades' are non-trivial tasks that involve many teams and stakeholders across
different business units or areas of responsibility including security.

Our current proposal is the creation of a Release SIG that will be responsible
for creating a schedule of 'epoch' releases for OpenTelemetry. These epoch
versions would essentially be a manifest pointing to a tested, documented, and
stable set of components that meet project stability requirements.

This is not a trivial undertaking, to be clear. It is also not as well-defined
as the earlier proposals in this blog -- those efforts will communicate many of
the requirements these epoch releases must follow, after all. To our maintainers
and contributors, though, a _non-goal_ of this effort is to force you to change

> your versioning or release cadence.

Reviewer 2: maybe reword to remove "non-goal", it read to me as a "non-goal but outcome of this effort ..."

These changes are setting up the next chapter of our success, and becoming truly
ubiquitous.

Our mission as a project is not changing, but our priorities are.

1. Stability and usability for all developers and users.
2. Clear packaging, installation, and usage paths.
3. Predictability and consistency.

For contributors and maintainers, what does this mean? We'll fast-track
proposals that align with these priorities. If there's feature work or
instrumentation that doesn't align to this, that's fine -- we'd ask that you
work on it outside the project and discover where our existing integration
points and patterns don't work; That's good feedback, and will help us improve
the specification for everyone.

For maintainers, contributors, and integrators -- we would appreciate your
feedback in
[this GitHub Discussion](https://github.com/open-telemetry/community/discussions/3098)
on the topics and proposals raised here. You can also send feedback on this
proposal to [feedback@opentelemetry.io](mailto:feedback@opentelemetry.io) or on
the CNCF Slack in the #opentelemetry channel. We also look forward to meeting

> the cloud native community in person at KubeCon this fall -- please join us

Reviewer 2: ```suggestion
the cloud native community in person at KubeCon next week -- please join us
```

---
title: Evolving OpenTelemetry's Stabilization and Release Practices
linkTitle: Stability Proposal Announcement # Mandatory, make sure that your short title.

> date: 2025-10-23 # Put the current date, we will keep the date updated until your PR is merged

Reviewer 13: ```suggestion
date: 2025-11-07
```

author:
  >- # If you have only one author, then add the single name on this line in quotes.
  "OpenTelemetry Governance Committee"
# canonical_url: http://somewhere.else/ # TODO: if this blog post has been posted somewhere else already, uncomment & provide the canonical URL here.
issue: the issue ID for this blog post # TODO: See https://opentelemetry.io/docs/contributing/blog/ for details
sig: Governance Committee # TODO: add the name of the SIG that sponsors this blog post
cSpell:ignore: deprioritize incentivized rollouts
---

## Summary

OpenTelemetry is, by any metric, one of the largest and most exciting projects
in the cloud native space. Over the past five years, this community has come
together to build one of the most essential observability projects in history.
We're not resting on our laurels, though. The project consistently seeks out,

> and listens, to feedback from a wide array of stakeholders. What we're hearing

Reviewer 14: ```suggestion
and listens to, feedback from a wide array of stakeholders. What we're hearing
```

deployment and upgrade of both SDKs and Collectors.

## Looking Forward

These changes are a reflection of the impact and importance of OpenTelemetry to
the cloud native software community.
[OpenTelemetry has been the second highest velocity project in the CNCF over the past few years](https://www.cncf.io/wp-content/uploads/2025/04/CNCF-Annual-Report-2024_v2.pdf),
and
[nearly 50% of surveyed cloud native end user companies have adopted the project](https://www.cncf.io/wp-content/uploads/2025/04/cncf_annual_survey24_031225a.pdf).
These changes are setting up the next chapter of our success, and becoming truly
ubiquitous.

Our mission as a project is not changing, but our priorities are.

1. Stability and usability for all developers and users.
2. Clear packaging, installation, and usage paths.
3. Predictability and consistency.

For contributors and maintainers, what does this mean? We'll fast-track
proposals that align with these priorities. If there's feature work or
instrumentation that doesn't align to this, that's fine -- we'd ask that you
work on it outside the project and discover where our existing integration

> points and patterns don't work; That's good feedback, and will help us improve

Reviewer 14: ```suggestion
points and patterns don't work. That's good feedback, and will help us improve
```
